### PR TITLE
restrict second argument of `size` and `axes` to Integer for most types

### PR DESCRIPTION
**Pull request by JeffBezanson**
This argument had a variety of strange behaviors:
```
julia> size([1], Int8(1))
ERROR: TypeError: in arraysize, expected Int64, got Int8

julia> size((), Int8(1))
0

julia> size((), :x)
ERROR: ArgumentError: invalid tuple dimension x

julia> size((1,), 1.0)
1

julia> size(1, :x)
ERROR: MethodError: Cannot `convert` an object of type Symbol to an object of type Int64
```

After:
```
julia> size([1], Int8(1))
1

julia> size((), Int8(1))
0

julia> size((), :x)
ERROR: MethodError: no method matching size(::Tuple{}, ::Symbol)

julia> size((1,), 1.0)
ERROR: MethodError: no method matching size(::Tuple{Int64}, ::Float64)

julia> size(1, :x)
ERROR: MethodError: no method matching size(::Int64, ::Symbol)
```

### CODE REVIEW

File: base/strings/util.jl [hunk 1]
```diff
@@ -602,7 +602,7 @@ function ascii(s::String)
     end
     return s
 end
-@noinline __throw_invalid_ascii(s, i) = throw(ArgumentError("invalid ASCII at index $i in $(repr(s))"))
+@noinline __throw_invalid_ascii(s::String, i::Int) = throw(ArgumentError("invalid ASCII at index $i in $(repr(s))"))
```
Comment (staticfloat) [inline base/strings/util.jl]: This seems like a mostly-unrelated change; is this just cleanup?  Do the type arguments help codegen because of `@noinline`?

Comment (JeffBezanson) [reply base/strings/util.jl]: Yes, it's unrelated. Just good style to write the types since we know them exactly.

**Comment by JeffBezanson** [comment]
Ok, this shouldn't be done for the `AbstractArray` fallback since it will introduce new ambiguities. Changed to a typeassert instead.

**Review by mbauman — APPROVED** [review]
